viz/LaserLine.cpp

```diff
@@ -2,44 +2,158 @@
 
 LaserLine::LaserLine()
-    : p(new Data)
 {
+    color = osg::Vec4(0.0, 1.0, 0.0, 0.0);
+    vertices = new osg::Vec3Array();
+    lineHolder = new LineHolder();
+    showAllLines = false;
 }
 
 LaserLine::~LaserLine()
 {
-    delete p;
 }
 
 osg::ref_ptr<osg::Node> LaserLine::createMainNode()
 {
-    // Geode is a common node used for vizkit3d plugins. It allows to display
-    // "arbitrary" geometries
-    return new osg::Geode();
+    return lineHolder->getRootNode();
 }
 
-void LaserLine::updateMainNode ( osg::Node* node )
+void LaserLine::updateMainNode (osg::Node* node)
 {
-    osg::Geode* geode = static_cast<osg::Geode*>(node);
-    // Update the main node using the data in p->data
+    lineHolder->update(vertices, color);
 }
 
 void LaserLine::updateDataIntern(base::samples::LaserScan const& value)
 {
-    p->data = value;
-    std::cout << "got new sample data" << std::endl;
+    loadLineVerticesFromLaserScan(value);
 }
 
+bool LaserLine::isShowRange() const
+{
+    return lineHolder->isShowRange();
+}
+void LaserLine::setShowRange(bool value)
+{
+    lineHolder->setShowRange(value);
+}
+
+QColor LaserLine::getColor()
+{
+    QColor qcolor;
+    qcolor.setRgbF(color[0], color[1], color[2], color[3]);
+    return qcolor;
+}
+
+void LaserLine::setColor(QColor value)
+{
+    if (value.isValid()) {
+        color.set(value.redF(),
+                  value.greenF(),
+                  value.blueF(),
+                  value.alphaF());
+    }
+}
+
+double LaserLine::getLineWidth()
+{
+    return (double)lineHolder->getLineWidth();
+}
+
+void LaserLine::setLineWidth(double lineWidth)
+{
+    lineHolder->setLineWidth((float)lineWidth);
+}
+
+bool LaserLine::getShowAllLines()
+{
+    return showAllLines;
+}
+
+void LaserLine::setShowAllLines(bool value)
+{
+    showAllLines = value;
+}
+
+osg::ref_ptr<osg::Vec3Array> LaserLine::computePoints(base::samples::LaserScan const& laserScan, std::vector<bool>& validity)
+{
+    osg::Vec3Array *points = new osg::Vec3Array();
+
+    std::vector<uint32_t> ranges = laserScan.ranges;
+
+    double angle = laserScan.start_angle; //start angle
+    double angleStep = laserScan.angular_resolution; //angle step
+
+    //convert max range to Point and convert the value from meters to milimiters
+    Eigen::Vector3d maxPt = Eigen::Vector3d(meters_to_milimeters(laserScan.maxRange), 0.0, 0.0);
+
+    for (int i = 0; i < ranges.size(); i++) {
+
+        //compute the angle rotation
+        Eigen::Quaterniond rot = Eigen::Quaterniond(Eigen::AngleAxisd(angle + (i * angleStep), Eigen::Vector3d::UnitZ()));
+        Eigen::Vector3d pt =  rot * Eigen::Vector3d(meters_to_milimeters(ranges[i]), 0.0, 0.0);
+        Eigen::Vector3d limit =  rot * maxPt;
+
+        points->push_back(osg::Vec3f(pt[0], pt[1], pt[2]));
+
+        //check if range is valid and if the point is less than the limit
+        validity.push_back(laserScan.isRangeValid(ranges[i]) && (pt[0] < limit[0] && pt[1] < limit[1]));
```

Comment:
I still don't understand the comparison of pt and limit. As I understand it, it is equivalent to

~~~
if (pt.x < limit.x && pt.y < limit.y)
~~~

Am I right ?